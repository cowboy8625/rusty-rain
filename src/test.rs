use super::{cli::Cli, Parser, Rain};
use std::fmt::Write;

struct SnapshotOptions {
    label: String,
    cycles: usize,
    width: usize,
    height: usize,
    direction: super::Direction,
    chars: super::Characters,
}

impl Default for SnapshotOptions {
    fn default() -> Self {
        Self {
            label: String::new(),
            cycles: 25,
            width: 40,
            height: 20,
            direction: super::Direction::Down,
            chars: super::Characters::Bin,
        }
    }
}

fn display<const N: usize>(id: usize, window: &mut String, rain: &Rain<N>) {
    let width = rain.width;
    let height = rain.height;
    let char_width = rain.group.width() as usize;
    let id_str = format!("{:02X}", id);
    write!(
        window,
        "{:-^width$}\n",
        id_str,
        width = width * char_width + 5
    )
    .unwrap();
    for (i, chunk) in rain.screen_buffer.chunks(width).enumerate() {
        write!(window, "{:02X} |", i).unwrap();
        write!(
            window,
            "{}|",
            &chunk
                .iter()
                .map(|c| c.display(char_width))
                .collect::<String>()
        )
        .unwrap();
        if i == height {
            continue;
        }
        write!(window, "\n").unwrap();
    }
}

fn set_up_snapshot(options: SnapshotOptions) {
    let SnapshotOptions {
        label,
        cycles,
        width,
        height,
        chars,
        direction,
    } = options;
    let mut cli = Cli::parse();
    cli.chars = chars;
    cli.direction = direction;
    let mut rain = Rain::<1024>::new(width, height, &cli);
    let mut window = String::new();
    for id in 0..cycles {
        rain.update();
        rain.update_screen_buffer().unwrap();
        display(id, &mut window, &rain);

        std::thread::sleep(std::time::Duration::from_millis(100));
    }
    insta::assert_snapshot!(label, window);
}

macro_rules! snapshot {
    ($name:ident) => {
        #[test]
        fn $name() {
            set_up_snapshot(SnapshotOptions {
                label: stringify!($name).replace("test_", "").to_string(),
                ..Default::default()
            });
        }
    };
    ($name:ident, $cycles:expr,$width:expr, $height:expr, $direction:expr, $chars:expr) => {
        #[test]
        fn $name() {
            set_up_snapshot(SnapshotOptions {
                label: stringify!($name).replace("test_", "").to_string(),
                cycles: $cycles,
                width: $width,
                height: $height,
                direction: $direction,
                chars: $chars,
            });
        }
    };
}

snapshot!(test_screen_buffer);
snapshot!(
    test_screen_buffer_direction_right_emoji_moon_double_width,
    50,
    32,
    10,
    super::Direction::Right,
    super::Characters::Moon
);
snapshot!(
    test_screen_buffer_direction_left_emoji_crab_double_width,
    50,
    32,
    10,
    super::Direction::Left,
    super::Characters::Crab
);

#[test]
fn test_gen_shade_color() {
    use super::{gen_shade_color, Color};
    use pretty_assertions::assert_eq;
    let bc = Color::Rgb { r: 0, g: 255, b: 0 };
    let length = 10;

    let colors = gen_shade_color(bc, length);

    assert_eq!(colors.len(), length as usize);
    assert_eq!(colors.first(), Some(&Color::Rgb { r: 0, g: 225, b: 0 }));
    assert_eq!(colors.last(), Some(&Color::Rgb { r: 0, g: 0, b: 0 }));
}
