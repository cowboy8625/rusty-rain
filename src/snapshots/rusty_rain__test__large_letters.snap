---
source: src/test.rs
expression: window
---
-----------------00------------------
00 |                                |
01 |                                |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------01------------------
00 |                                |
01 |                                |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------02------------------
00 |                  Ｇ            |
01 |                                |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------03------------------
00 |                  Ｇ            |
01 |                                |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------04------------------
00 |                  Ｇ        Ｙ  |
01 |                  Ｉ            |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------05------------------
00 |                  Ｇ      ＢＹ  |
01 |                  Ｉ            |
02 |                                |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------06------------------
00 |            Ｒ    Ｇ      ＢＹ  |
01 |                  Ｉ        Ｊ  |
02 |                  Ｒ            |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------07------------------
00 |            Ｒ  ＵＧ      ＢＹ  |
01 |            Ｇ    Ｉ      ＦＪ  |
02 |                  Ｒ            |
03 |                                |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------08------------------
00 |Ｕ          Ｒ  ＵＧ  Ｙ  ＢＹ  |
01 |            Ｇ  ＯＩ      ＦＪ  |
02 |            Ｖ    Ｒ        Ｗ  |
03 |                  Ａ            |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------09------------------
00 |Ｕ          Ｒ  ＵＧ  Ｙ  ＢＹ  |
01 |            Ｇ  ＯＩ      ＦＪ  |
02 |            Ｖ  ＦＲ      ＩＷ  |
03 |            Ａ    Ａ            |
04 |                                |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------0A------------------
00 |Ｕ          Ｒ  ＵＧ  Ｙ  ＢＹ  |
01 |Ｔ          Ｇ  ＯＩ  Ｎ  ＦＪ  |
02 |            Ｖ  ＦＲ      ＩＷ  |
03 |            Ａ  ＴＡ        Ｆ  |
04 |            Ｊ    Ｃ            |
05 |                                |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------0B------------------
00 |ＵＧ              Ｇ  Ｙ  ＢＹ  |
01 |Ｔ          Ｇ  ＯＩ  Ｎ  ＦＪ  |
02 |            Ｖ  ＦＲ      ＩＷ  |
03 |            Ａ  ＴＡ      ＶＦ  |
04 |            Ｊ  ＴＣ            |
05 |            Ｘ                  |
06 |                                |
07 |                                |
08 |                                |
09 |                                |
-----------------0C------------------
00 |ＵＧ          Ｐ      Ｙ  ＢＹ  |
01 |Ｔ                Ｉ  Ｎ  ＦＪ  |
02 |Ｍ          Ｖ  ＦＲ  Ｄ  ＩＷ  |
03 |            Ａ  ＴＡ      ＶＦ  |
04 |            Ｊ  ＴＣ        Ｘ  |
05 |            Ｘ  ＱＬ            |
06 |            Ｒ                  |
07 |                                |
08 |                                |
09 |                                |
-----------------0D------------------
00 |ＵＧ      Ｌ  Ｐ      Ｙ  ＢＹ  |
01 |ＴＬ          Ｗ  Ｉ  Ｎ  ＦＪ  |
02 |Ｍ                Ｒ  Ｄ  ＩＷ  |
03 |            Ａ  ＴＡ      ＶＦ  |
04 |            Ｊ  ＴＣ      ＬＸ  |
05 |            Ｘ  ＱＬ            |
06 |            Ｒ  Ｒ              |
07 |            Ｙ                  |
08 |                                |
09 |                                |
-----------------0E------------------
00 |ＵＧ      Ｌ  Ｐ    ＧＹＫＢ    |
01 |ＴＬ          Ｗ      Ｎ  ＦＪ  |
02 |Ｍ            Ｌ  Ｒ  Ｄ  ＩＷ  |
03 |Ｍ                Ａ  Ｊ  ＶＦ  |
04 |            Ｊ  ＴＣ      ＬＸ  |
05 |            Ｘ  ＱＬ        Ｇ  |
06 |            Ｒ  ＲＹ            |
07 |            Ｙ  Ｇ              |
08 |            Ｂ                  |
09 |                                |
-----------------0F------------------
00 |ＵＧＲＥＥＬ  Ｐ    ＧＹＫ      |
01 |ＴＬ      Ｃ  Ｗ    ＬＮ  ＦＪ  |
02 |ＭＫ          Ｌ  Ｒ  Ｄ  ＩＷ  |
03 |Ｍ            Ｇ  Ａ  Ｊ  ＶＦ  |
04 |                  Ｃ      ＬＸ  |
05 |            Ｘ  ＱＬ      ＶＧ  |
06 |            Ｒ  ＲＹ            |
07 |            Ｙ  Ｇ              |
08 |            Ｂ  Ｌ              |
09 |            Ｘ                  |
-----------------10------------------
00 |ＵＧＲＥＥＬ        ＧＹＫ      |
01 |ＴＬ      Ｃ  Ｗ    ＬＮＡＦ    |
02 |ＭＫ          Ｌ    ＣＤ  ＩＷ  |
03 |Ｍ            Ｇ  Ａ  Ｊ  ＶＦ  |
04 |Ｍ            Ａ  Ｃ  Ｒ  ＬＸ  |
05 |                  Ｌ      ＶＧ  |
06 |            Ｒ  ＲＹ        Ｗ  |
07 |            Ｙ  ＧＧ            |
08 |            Ｂ  Ｌ              |
09 |            Ｘ  Ｖ              |
-----------------11------------------
00 |ＵＧＲＥＥＬ        ＧＹＫ      |
01 |ＴＬＲＮＧＣ        ＬＮＡ      |
02 |ＭＫ      Ｌ  Ｌ    ＣＤ  ＩＷ  |
03 |ＭＵ          Ｇ  ＡＬＪ  ＶＦ  |
04 |Ｍ            Ａ  Ｃ  Ｒ  ＬＸ  |
05 |              Ｖ  Ｌ      ＶＧ  |
06 |                  Ｙ      ＩＷ  |
07 |            Ｙ  ＧＧ            |
08 |            Ｂ  Ｌ              |
09 |            Ｘ  Ｖ              |
-----------------12------------------
00 |  ＧＲＥＥＬ        Ｇ  Ｋ    Ｐ|
01 |ＴＬＲＮＧＣ        ＬＮＡ      |
02 |ＭＫ      Ｌ        ＣＤＹＩ    |
03 |ＭＵ          Ｇ    ＬＪ  ＶＦ  |
04 |Ｍ            Ａ  ＣＡＲ  ＬＸ  |
05 |Ｖ            Ｖ  Ｌ  Ａ  ＶＧ  |
06 |              Ｄ  Ｙ      ＩＷ  |
07 |                  Ｇ        Ｐ  |
08 |            Ｂ  ＬＧ            |
09 |            Ｘ  Ｖ              |
-----------------13------------------
00 |  ＧＲＥＥＬ            Ｋ    Ｐ|
01 |ＴＬＲＮＧＣ        ＬＮＡ    Ｒ|
02 |ＭＫＹＴＹＬ        ＣＤＹ      |
03 |ＭＵ      Ａ        ＬＪ  ＶＦ  |
04 |ＭＢ          Ａ  ＣＡＲ  ＬＸ  |
05 |Ｖ            Ｖ  ＬＣＡ  ＶＧ  |
06 |              Ｄ  Ｙ      ＩＷ  |
07 |              Ｇ  Ｇ      ＫＰ  |
08 |                  Ｇ            |
09 |            Ｘ  Ｖ              |
-----------------14------------------
00 |  ＧＲＥＥＬ            Ｋ    Ｐ|
01 |  ＬＲＮＧＣ            Ａ    Ｒ|
02 |ＭＫＹＴＹＬ        ＣＤＹ    Ｈ|
03 |ＭＵ      Ａ        ＬＪＴＶ    |
04 |ＭＢ                ＡＲ  ＬＸ  |
05 |Ｖ            Ｖ  ＬＣＡ  ＶＧ  |
06 |Ｊ            Ｄ  ＹＤＴ  ＩＷ  |
07 |              Ｇ  Ｇ      ＫＰ  |
08 |              Ｊ  Ｇ        Ｈ  |
09 |                  Ｅ            |
-----------------15------------------
00 |    ＲＥＥＬ            Ｋ    Ｐ|
01 |  ＬＲＮＧＣ            Ａ    Ｒ|
02 |ＭＫＹＴＹＬ          ＤＹ    Ｈ|
03 |ＭＵＸＯＫＡ        ＬＪＴ    Ｄ|
04 |ＭＢ      Ｃ        ＡＲ  ＬＸ  |
05 |ＶＵ              ＬＣＡ  ＶＧ  |
06 |Ｊ            Ｄ  ＹＤＴ  ＩＷ  |
07 |              Ｇ  ＧＲ    ＫＰ  |
08 |              Ｊ  Ｇ      ＶＨ  |
09 |              Ａ  Ｅ            |
-----------------16------------------
00 |    ＲＥＥＬ            Ｋ      |
01 |  ＬＲＮＧＣ            Ａ    Ｒ|
02 |  ＫＹＴＹＬ            Ｙ    Ｈ|
03 |ＭＵＸＯＫＡ          ＪＴ    Ｄ|
04 |ＭＢ      Ｃ        ＡＲＰＬ  Ｓ|
05 |ＶＵ                ＣＡ  ＶＧ  |
06 |Ｊ                ＹＤＴ  ＩＷ  |
07 |Ｄ            Ｇ  ＧＲＴ  ＫＰ  |
08 |              Ｊ  ＧＱ    ＶＨ  |
09 |              Ａ  Ｅ        Ｌ  |
-----------------17------------------
00 |    Ｒ  Ｅ              Ｋ      |
01 |    ＲＮＧＣ            Ａ      |
02 |  ＫＹＴＹＬ            Ｙ    Ｈ|
03 |ＭＵＸＯＫＡ          ＪＴ    Ｄ|
04 |ＭＢＧＮＳＣ          ＲＰ    Ｓ|
05 |ＶＵ      Ｄ        ＣＡ  ＶＧＮ|
06 |ＪＯ              ＹＤＴ  ＩＷ  |
07 |Ｄ                ＧＲＴ  ＫＰ  |
08 |              Ｊ  ＧＱ    ＶＨ  |
09 |              Ａ  ＥＨ    ＪＬ  |
-----------------18------------------
00 |    Ｒ  Ｅ  Ｃ  Ｌ              |
01 |    ＲＮＧＣ            Ａ      |
02 |  ＫＹＴＹＬ            Ｙ      |
03 |  ＵＸＯＫＡ            Ｔ    Ｄ|
04 |ＭＢＧＮＳＣ          ＲＰ    Ｓ|
05 |ＶＵ      Ｄ          ＡＷＶ  Ｎ|
06 |ＪＯ                ＤＴ  ＩＷＧ|
07 |Ｄ                ＧＲＴ  ＫＰ  |
08 |Ｌ                ＧＱＶ  ＶＨ  |
09 |              Ａ  ＥＨ    ＪＬ  |
-----------------19------------------
00 |            Ｃ  Ｌ              |
01 |    Ｒ  Ｇ              Ａ      |
02 |    ＹＴＹＬ            Ｙ      |
03 |  ＵＸＯＫＡ            Ｔ      |
04 |ＭＢＧＮＳＣ          ＲＰ    Ｓ|
05 |ＶＵＣＨＴＤ          ＡＷ    Ｎ|
06 |ＪＯ      Ｒ          Ｔ  ＩＷＧ|
07 |ＤＦ              ＧＲＴ  ＫＰＢ|
08 |Ｌ                ＧＱＶ  ＶＨ  |
09 |                  ＥＨ    ＪＬ  |
-----------------1A------------------
00 |            Ｃ  Ｌ              |
01 |    Ｒ  Ｇ  Ｋ  Ｋ              |
02 |    ＹＴＹＬ            Ｙ      |
03 |  ＵＸＯＫＡ            Ｔ      |
04 |  ＢＧＮＳＣ            Ｐ      |
05 |ＶＵＣＨＴＤ          ＡＷ    Ｎ|
06 |ＪＯ      Ｒ          ＴＬＩ  Ｇ|
07 |ＤＦ                  Ｔ  ＫＰＢ|
08 |Ｌ                ＧＱＶ  ＶＨＳ|
09 |Ｌ                ＥＨＮ  ＪＬ  |
-----------------1B------------------
00 |            Ｃ  Ｌ              |
01 |            Ｋ  Ｋ              |
02 |    Ｙ  Ｙ              Ｙ      |
03 |    ＸＯＫＡ            Ｔ      |
04 |  ＢＧＮＳＣ            Ｐ      |
05 |ＶＵＣＨＴＤ          ＡＷ      |
06 |ＪＯＤＧＣＲ          ＴＬ    Ｇ|
07 |ＤＦ      Ｑ          Ｔ  ＫＰＢ|
08 |ＬＴ              Ｇ  Ｖ  ＶＨＳ|
09 |Ｌ                ＥＨＮ  ＪＬＷ|
-----------------1C------------------
00 |            Ｃ  Ｌ              |
01 |            Ｋ  Ｋ              |
02 |    Ｙ  Ｙ  Ａ  Ｕ              |
03 |    ＸＯＫＡ            Ｔ      |
04 |  ＢＧＮＳＣ            Ｐ      |
05 |  ＵＣＨＴＤ            Ｗ      |
06 |ＪＯＤＧＣＲ          ＴＬ      |
07 |ＤＦ      Ｑ          ＴＧＫ  Ｂ|
08 |ＬＴ                  Ｖ  ＶＨＳ|
09 |Ｌ                Ｅ  Ｎ  ＪＬＷ|
-----------------1D------------------
00 |            ＣＷＬ              |
01 |            Ｋ  Ｋ              |
02 |            Ａ  Ｕ              |
03 |    Ｘ  Ｋ              Ｔ      |
04 |    ＧＮＳＣ            Ｐ      |
05 |  ＵＣＨＴＤ            Ｗ      |
06 |ＪＯＤＧＣＲ          ＴＬ      |
07 |ＤＦＶＹＶＱ          ＴＧ      |
08 |ＬＴ      Ｈ          Ｖ  ＶＨＳ|
09 |ＬＴ              Ｅ  Ｎ  ＪＬＷ|
-----------------1E------------------
00 |            ＣＷＬ              |
01 |            Ｋ  Ｋ              |
02 |            Ａ  Ｕ              |
03 |    Ｘ  Ｋ  Ｙ  Ｂ              |
04 |    ＧＮＳＣ            Ｐ      |
05 |  ＵＣＨＴＤ            Ｗ      |
06 |  ＯＤＧＣＲ            Ｌ      |
07 |ＤＦＶＹＶＱ          ＴＧ      |
08 |ＬＴ      Ｈ          ＶＡＶ    |
09 |ＬＴ                  Ｎ  ＪＬＷ|
-----------------1F------------------
00 |            ＣＷＬ              |
01 |            ＫＴＫ              |
02 |            Ａ  Ｕ              |
03 |            Ｙ  Ｂ              |
04 |    Ｇ  Ｓ              Ｐ      |
05 |    ＣＨＴＤ            Ｗ      |
06 |  ＯＤＧＣＲ            Ｌ      |
07 |ＤＦＶＹＶＱ          ＴＧ      |
08 |ＬＴＦＪＦＨ          ＶＡ      |
09 |ＬＴ      Ｈ          Ｎ  ＪＬ  |
-----------------20------------------
00 |              ＷＬ  Ｑ          |
01 |            ＫＴＫ              |
02 |            Ａ  Ｕ              |
03 |            Ｙ  Ｂ              |
04 |    Ｇ  Ｓ  Ｔ  Ｕ              |
05 |    ＣＨＴＤ            Ｗ      |
06 |  ＯＤＧＣＲ            Ｌ      |
07 |  ＦＶＹＶＱ            Ｇ      |
08 |ＬＴＦＪＦＨ          ＶＡ      |
09 |ＬＴ      Ｈ          ＮＶＪ    |
-----------------21------------------
00 |              ＷＬ  Ｑ          |
01 |            ＫＴＫ              |
02 |            ＡＪＵ              |
03 |            Ｙ  Ｂ              |
04 |            Ｔ  Ｕ              |
05 |    Ｃ  Ｔ              Ｗ      |
06 |    ＤＧＣＲ            Ｌ      |
07 |  ＦＶＹＶＱ            Ｇ      |
08 |ＬＴＦＪＦＨ          ＶＡ      |
09 |ＬＴＯＪＧＨ          ＮＶ      |
-----------------22------------------
00 |              Ｗ    Ｑ        Ｐ|
01 |              ＴＫ  Ｂ          |
02 |            ＡＪＵ              |
03 |            Ｙ  Ｂ              |
04 |            Ｔ  Ｕ              |
05 |    Ｃ  Ｔ  Ｐ  Ｏ              |
06 |    ＤＧＣＲ            Ｌ      |
07 |  ＦＶＹＶＱ            Ｇ      |
08 |  ＴＦＪＦＨ            Ａ      |
09 |ＬＴＯＪＧＨ          ＮＶ      |
-----------------23------------------
00 |              Ｗ    Ｑ        Ｐ|
01 |              ＴＫ  Ｂ        Ｓ|
02 |            ＡＪＵ              |
03 |            ＹＣＢ              |
04 |            Ｔ  Ｕ              |
05 |            Ｐ  Ｏ              |
06 |    Ｄ  Ｃ              Ｌ      |
07 |    ＶＹＶＱ            Ｇ      |
08 |  ＴＦＪＦＨ            Ａ      |
09 |ＬＴＯＪＧＨ          ＮＶ      |
-----------------24------------------
00 |              Ｗ  ＡＱ        Ｐ|
01 |              Ｔ    Ｂ        Ｓ|
02 |              ＪＵ  Ａ        Ｙ|
03 |            ＹＣＢ              |
04 |            Ｔ  Ｕ              |
05 |            Ｐ  Ｏ              |
06 |    Ｄ  Ｃ  Ｗ  Ｆ              |
07 |    ＶＹＶＱ            Ｇ      |
08 |  ＴＦＪＦＨ            Ａ      |
09 |  ＴＯＪＧＨ            Ｖ      |
-----------------25------------------
00 |              Ｗ  ＡＱ        Ｐ|
01 |              Ｔ    Ｂ        Ｓ|
02 |              ＪＵ  Ａ        Ｙ|
03 |            ＹＣＢ            Ｃ|
04 |            ＴＭＵ              |
05 |            Ｐ  Ｏ              |
06 |            Ｗ  Ｆ              |
07 |    Ｖ  Ｖ              Ｇ      |
08 |    ＦＪＦＨ            Ａ      |
09 |  ＴＯＪＧＨ            Ｖ      |
-----------------26------------------
00 |              Ｗ  ＡＱ      ＳＰ|
01 |              Ｔ  ＬＢ        Ｓ|
02 |              Ｊ    Ａ        Ｙ|
03 |              ＣＢ  Ｙ        Ｃ|
04 |            ＴＭＵ            Ｅ|
05 |            Ｐ  Ｏ              |
06 |            Ｗ  Ｆ              |
07 |    Ｖ  Ｖ  Ｌ  Ｔ              |
08 |    ＦＪＦＨ            Ａ      |
09 |  ＴＯＪＧＨ            Ｖ      |
-----------------27------------------
00 |                  ＡＱ    ＷＳ  |
01 |              Ｔ  ＬＢ        Ｓ|
02 |              Ｊ    Ａ        Ｙ|
03 |              ＣＢ  Ｙ        Ｃ|
04 |            ＴＭＵ            Ｅ|
05 |            ＰＤＯ            Ｑ|
06 |            Ｗ  Ｆ              |
07 |            Ｌ  Ｔ              |
08 |    Ｆ  Ｆ              Ａ      |
09 |    ＯＪＧＨ            Ｖ      |
-----------------28------------------
00 |                  Ａ      ＷＳ  |
01 |              Ｔ  ＬＢ      Ｔ  |
02 |              Ｊ  ＶＡ        Ｙ|
03 |              Ｃ    Ｙ        Ｃ|
04 |              ＭＵ  Ｙ        Ｅ|
05 |            ＰＤＯ            Ｑ|
06 |            Ｗ  Ｆ            Ｑ|
07 |            Ｌ  Ｔ              |
08 |    Ｆ  Ｆ  Ｇ  Ｔ              |
09 |    ＯＪＧＨ            Ｖ      |
-----------------29------------------
00 |                  Ａ  Ｒ  ＷＳ  |
01 |                  ＬＢ    ＱＴ  |
02 |              Ｊ  ＶＡ          |
03 |              Ｃ    Ｙ        Ｃ|
04 |              ＭＵ  Ｙ        Ｅ|
05 |            ＰＤＯ            Ｑ|
06 |            ＷＣＦ            Ｑ|
07 |            Ｌ  Ｔ            Ｅ|
08 |            Ｇ  Ｔ              |
09 |    Ｏ  Ｇ              Ｖ      |
-----------------2A------------------
00 |Ｅ                Ａ  Ｒ  ＷＳ  |
01 |                  Ｌ  Ｓ  ＱＴ  |
02 |              Ｊ  ＶＡ      Ｈ  |
03 |              Ｃ  ＱＹ          |
04 |              Ｍ    Ｙ        Ｅ|
05 |              ＤＯ  Ｎ        Ｑ|
06 |            ＷＣＦ            Ｑ|
07 |            Ｌ  Ｔ            Ｅ|
08 |            Ｇ  Ｔ            Ｎ|
09 |    Ｏ  Ｇ  Ａ  Ｑ              |
-----------------2B------------------
00 |Ｅ                Ａ  Ｒ  ＷＳ  |
01 |                  Ｌ  Ｓ  ＱＴ  |
02 |                  ＶＡＱ  ＲＨ  |
03 |              Ｃ  ＱＹ          |
04 |              Ｍ    Ｙ          |
05 |              ＤＯ  Ｎ        Ｑ|
06 |            ＷＣＦ            Ｑ|
07 |            ＬＭＴ            Ｅ|
08 |            Ｇ  Ｔ            Ｎ|
09 |            Ａ  Ｑ            Ｔ|
-----------------2C------------------
00 |ＥＣ                  Ｒ  ＷＳ  |
01 |Ｔ                Ｌ  Ｓ  ＱＴ  |
02 |                  Ｖ  Ｑ  ＲＨ  |
03 |              Ｃ  ＱＹＢ    Ｕ  |
04 |              Ｍ  ＧＹ          |
05 |              Ｄ    Ｎ          |
06 |              ＣＦ  Ｄ        Ｑ|
07 |            ＬＭＴ            Ｅ|
08 |            Ｇ  Ｔ            Ｎ|
09 |            Ａ  Ｑ            Ｔ|
-----------------2D------------------
00 |ＥＣ                      ＷＳ  |
01 |ＴＮ              Ｌ  Ｓ  ＱＴ  |
02 |                  Ｖ  Ｑ  ＲＨ  |
03 |                  ＱＹＢ  ＢＵ  |
04 |              Ｍ  ＧＹＪ        |
05 |              Ｄ    Ｎ          |
06 |              ＣＦ  Ｄ          |
07 |            ＬＭＴ            Ｅ|
08 |            ＧＢＴ            Ｎ|
09 |            Ａ  Ｑ            Ｔ|
-----------------2E------------------
00 |ＥＣ  Ｇ                  ＷＳ  |
01 |ＴＮ                      ＱＴ  |
02 |ＤＩ              Ｖ  Ｑ  ＲＨ  |
03 |                  Ｑ  Ｂ  ＢＵ  |
04 |              Ｍ  ＧＹＪ    Ｃ  |
05 |              Ｄ  ＷＮＤ        |
06 |              Ｃ    Ｄ          |
07 |              ＭＴ  Ｊ          |
08 |            ＧＢＴ            Ｎ|
09 |            Ａ  Ｑ            Ｔ|
-----------------2F------------------
00 |ＥＣ  Ｇ  Ｎ            ＤＷＳ  |
01 |ＴＮ  Ｙ                  ＱＴ  |
02 |ＤＩ              Ｖ      ＲＨ  |
03 |  Ｔ              Ｑ  Ｂ  ＢＵ  |
04 |                  ＧＹＪ  ＨＣ  |
05 |              Ｄ  ＷＮＤ        |
06 |              Ｃ    ＤＬ        |
07 |              ＭＴ  Ｊ          |
08 |            ＧＢＴ              |
09 |            ＡＱＱ            Ｔ|
-----------------30------------------
00 |Ｅ    ＧＴＮ            ＤＷ    |
01 |ＴＮ  Ｙ                ＲＱＴ  |
02 |ＤＩ  Ｗ                  ＲＨ  |
03 |ＮＴ              Ｑ      ＢＵ  |
04 |  Ｇ              Ｇ  Ｊ  ＨＣ  |
05 |              Ｄ  ＷＮＤ    Ｊ  |
06 |              Ｃ  ＴＤＬ        |
07 |              Ｍ    ＪＲ        |
08 |              ＢＴ  Ｒ          |
09 |            ＡＱＱ              |
-----------------31------------------
00 |Ｅ  ＢＧＴＮ            Ｄ      |
01 |Ｔ    ＹＹＧ            ＲＱＴ  |
02 |ＤＩ  Ｗ                ＫＲＨ  |
03 |ＮＴ  Ｐ          Ｑ      ＢＵ  |
04 |  Ｇ              Ｇ      ＨＣ  |
05 |  Ｈ              ＷＮＤ  ＯＪ  |
06 |              Ｃ  ＴＤＬ        |
07 |              Ｍ    ＪＲ        |
08 |              ＢＴ  ＲＨ        |
09 |            ＡＱＱ              |
